presto-main/src/main/java/com/facebook/presto/execution/FutureStateChange.java

```diff
@@ -66,14 +67,44 @@ public void complete(T newState, Executor executor)
     private void fireStateChange(T newState, Executor executor)
     {
         requireNonNull(executor, "executor is null");
-        Set<SettableFuture<T>> futures;
+        Set<StateTrackingFuture<T>> futures;
         synchronized (listeners) {
             futures = ImmutableSet.copyOf(listeners);
             listeners.clear();
         }
 
-        for (SettableFuture<T> future : futures) {
+        for (StateTrackingFuture<T> future : futures) {
             executor.execute(() -> future.set(newState));
         }
     }
+
+    /**
+     * Future cancellation can be expensive, because when a future is canceled
+     * a CancellationException is created and thrown under the hood. However, for
+     * tracking the state changes we really don't care about the CancellationExceptions
+     * and we only care about completing these futures. This future implementation
+     * just completes the future when it's canceled and gets rid of the cancellation
+     * overhead.
+     */
+    private static final class StateTrackingFuture<V>
+            extends AbstractFuture<V>
+    {
+        public static <V> StateTrackingFuture<V> create()
+        {
+            return new StateTrackingFuture<V>();
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning)
+        {
+            set(null);
```

Comment:
I am concerned that this changes the behavior of the future returned from `createNewListener` where a callback could see a null value.  Instead, how about we create a constant `CancellationException` and then `setException(constant)`?  Would that give us the same benefit?

Reply:
Yep it helps too.